### PR TITLE
explicitly set the API server to a random port

### PR DESCRIPTION
**Pull request by BenTheElder**
/hold
cc @Katharine

### CODE REVIEW

File: pkg/cluster/node.go [hunk 1]
```diff
@@ -65,9 +78,9 @@ func createNode(name, image, clusterLabel string) (handle *nodeHandle, err error
 			"--name", name, // ... and set the container name
 			// label the node with the cluster ID
 			"--label", clusterLabel,
-			"--expose", "6443", // expose API server port
-			// pick a random ephemeral port to forward to the API server
-			"--publish-all",
+			// publish selected port for the API server
+			"--expose", fmt.Sprintf("%d", port),
+			"-p", fmt.Sprintf("%d:%d", port, port),
```
Comment (munnerz) [inline pkg/cluster/node.go]: If we keep this as %d:6443, we can avoid the config template change I think? Not sure what the original motivations for the PR are

File: pkg/cluster/kubeadm/config.go [hunk 2]
```diff
@@ -56,7 +58,10 @@ const DefaultConfigTemplateAlphaV1orV2 = `# config generated by kind
 apiVersion: kubeadm.k8s.io/v1alpha2
 kind: MasterConfiguration
 kubernetesVersion: {{.KubernetesVersion}}
-# TODO(bentheelder): fix this upstream!
+clusterName: {{.ClusterName}}
```
Comment (munnerz) [inline pkg/cluster/kubeadm/config.go]: Do we need to set a cluster name? Is this related to setting the port?

File: pkg/cluster/kubeadm/config.go [hunk 3]
```diff
@@ -65,7 +70,6 @@ apiServerExtraVolumes:
   hostPath: /etc/nsswitch.conf
   writeable: false
   pathType: FileOrCreate
-clusterName: {{.ClusterName}}
```
Comment (BenTheElder) [inline pkg/cluster/kubeadm/config.go]: previously set here

File: pkg/cluster/node.go (hunk 1 above)
Comment (BenTheElder) [reply pkg/cluster/node.go]: making sure things consistently hit the correct port by informing everything what the api server port is.

**Comment by neolit123** [comment]
the InitConfiguration part LGTM

i will double check when 1.13 is near, but for v1beta1, and the current use case we can just do `DefaultConfigTemplateAlphaV3orBetaV1` it feels like.

File: pkg/cluster/node.go (hunk 1 above)
Comment (neolit123) [reply pkg/cluster/node.go]: i assume that the random port is motivated by security?

Comment (BenTheElder) [reply pkg/cluster/node.go]: actually because we support multiple clusters on one machine and we set localhost:port to point to them, we already set a random port today, we just don't use the same port within the container which makes some inconsistency with the api server port

**Comment by BenTheElder** [comment]
/hold cancel
I think we should actually do this. We're going to switch to kustomize to avoid issues with dowstream users setting the template, instead they will set a patch or patches to apply.

File: pkg/cluster/node.go [hunk 4]
```diff
@@ -83,9 +96,9 @@ func createNode(name, image, clusterLabel string) (handle *nodeHandle, err error
 		handle = &nodeHandle{name}
 	}
 	if err != nil {
-		return handle, errors.Wrap(err, "docker run error")
+		return handle, 0, errors.Wrap(err, "docker run error")
 	}
-	return handle, nil
+	return handle, port, nil
```
Comment (amwat) [inline pkg/cluster/node.go]: Add port to the nodeHandle struct?

Comment (BenTheElder) [reply pkg/cluster/node.go]: the port is totally orthogonal to the node handle.
